receiver/mongodbreceiver/extractor.go

```diff
@@ -0,0 +1,247 @@
+		mb.RecordMongodbGlobalLockTimeDataPoint(ts, waitTime)
+	}
+
+	// Collect Cache Hits & Misses
+	canCalculateCacheHits := true
+
+	cacheMisses, err := digForIntValue(document, []string{"wiredTiger", "cache", "pages read into cache"})
+	if err != nil {
+		e.logger.Error("failed to parse cache misses", zap.Error(err))
+		canCalculateCacheHits = false
+	} else {
+		mb.RecordMongodbCacheOperationsDataPoint(ts, cacheMisses, metadata.AttributeType.Miss)
+	}
+
+	totalCacheRequests, err := digForIntValue(document, []string{"wiredTiger", "cache", "pages requested from the cache"})
+	if err != nil {
+		e.logger.Debug("failed to parse total cache requests unable to calculate cache hits", zap.Error(err))
+		canCalculateCacheHits = false
+	}
+
+	if canCalculateCacheHits && totalCacheRequests > cacheMisses {
+		cacheHits := totalCacheRequests - cacheMisses
+		mb.RecordMongodbCacheOperationsDataPoint(ts, cacheHits, metadata.AttributeType.Hit)
+	}
+
+	// Collect Operations
+	for _, operation := range []string{
+		metadata.AttributeOperation.Insert,
+		metadata.AttributeOperation.Query,
+		metadata.AttributeOperation.Update,
+		metadata.AttributeOperation.Delete,
+		metadata.AttributeOperation.Getmore,
+		metadata.AttributeOperation.Command,
+	} {
+		count, err := digForIntValue(document, []string{"opcounters", operation})
+		if err != nil {
+			e.logger.Error("failed to parse operation", zap.Error(err), zap.String("operation", operation))
+			continue
+		}
+		mb.RecordMongodbOperationCountDataPoint(ts, count, operation)
+	}
+}
+
+func (e *extractor) extractGlobalLockWaitTime(document bson.M) (int64, error) {
+	var heldTimeUs int64
+
+	// Mongo version greater than or equal to 4.0 have it in the serverStats at "globalLock", "totalTime"
+	// reference: https://docs.mongodb.com/v4.0/reference/command/serverStatus/#server-status-global-lock
+	if e.version.GreaterThanOrEqual(Mongo40) {
+		heldTimeUs, _ = digForIntValue(document, []string{"globalLock", "totalTime"})
+	} else {
+		for _, lockType := range []string{"W", "R", "r", "w"} {
+			waitTime, err := digForIntValue(document, []string{"locks", ".", "timeAcquiringMicros", lockType})
+			if err == nil {
+				heldTimeUs += waitTime
+			}
+		}
+	}
+
+	if heldTimeUs != 0 {
+		htMilliseconds := heldTimeUs / 1000
+		return htMilliseconds, nil
+	}
+
+	e.logger.Warn("unable to find global lock time")
+	return 0, errors.New("was unable to calculate global lock time")
+}
+
+func digForIntValue(document bson.M, path []string) (int64, error) {
+	curItem, remainingPath := path[0], path[1:]
+	value := document[curItem]
+	if value == nil {
+		return 0, errors.New("nil found when digging for metric")
+	}
+
+	if len(remainingPath) == 0 {
+		switch v := value.(type) {
+		case int:
+			return int64(v), nil
+		case int32:
+			return int64(v), nil
+		case int64:
+			return v, nil
+		case float64:
+			return int64(v), nil
+		default:
+			return 0, fmt.Errorf("unexpected type found when parsing int: %v", reflect.TypeOf(value))
+		}
```

Comment:
Same question as elsewhere. Can we determine which of these are actually necessary?

Reply:
I've trimmed down the list to the ones I know are returned.